File: 107. Binary Tree Level Order Traversal II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> output = new ArrayList<List<Integer>>();
        if (root == null) {
            return output;
        }
        Queue<TreeNode> q1 = new LinkedList<>();
        Queue<TreeNode> q2 = new LinkedList<>();
        q1.add(root);
        while (!q1.isEmpty() || !q2.isEmpty()) {
            List<Integer> l1 = new ArrayList<Integer>();
            while (!q1.isEmpty()) {
                root = q1.poll();
                l1.add(root.val);
                if (root.left != null) {
                    q2.add(root.left);
                }
                if (root.right != null) {
                    q2.add(root.right);
                }
            }
            if (l1.size() > 0) {
                output.add(l1);
            }
            List<Integer> l2 = new ArrayList<Integer>();
            while (!q2.isEmpty()) {
                root = q2.poll();
                l2.add(root.val);
                if (root.left != null) {
                    q1.add(root.left);
                }
                if (root.right != null) {
                    q1.add(root.right);
                }
            }
            if (l2.size() > 0){
                output.add(l2);
            }
        }
       Collections.reverse(output);
       return output;
    }
}

File: 107. Binary Tree Level Order Traversal II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        
        def bfs(root):
            q1 = []
            q1.append(root)
            levels = []
            
            while q1:
                q2 = []
                levels.append([node.val for node in q1])
                
                while q1:
                    node = q1.pop(0)
                    
                    if node.left:
                        q2.append(node.left)
                    if node.right:
                        q2.append(node.right)
                
                q1 = q2
            
            return levels
        
        return(bfs(root)[::-1])
